Add verify module

lkatalin opened the pull request:
A series of small changes leading to the addition of the `verify` module to `attestation_types`.

The OpenSSL methods that are going to be checking signatures on various parts of the Quote are going to need a `Vec<u8>` or `&[u8]` representation of specific inner structures, so methods are added to get these back out from a `Quote`. Some additional helpful changes are made to the `SigData` struct (the last part of a Quote holding useful data for verifying the Quote).

The `verify` module has some functions (in `cert_chain,rs`, `sig.rs`, `key.rs`) we're not using currently but might be helpful generally. I could also delete these. (They are from last year's demo.) It was complaining about dead code, so there's probably some better way to structure it to make that go away. I'll clean all this up more later anyway but I don't mind making changes to get this merged.

In this PR:
- Adding methods on `ECDSAP256Sig` and `ECDSAPKey` to get a vector of the structs
- Adding a `PCKCertChain` type
- Adding the cert data's length as a field to the `SigData` struct
- Changing the cert data type that `SigData` expects because this is what's in the V3 Quote
- Fixing some SigData` methods (suggestions welcome)
- Adding a method on `SigData` to get the cert data as a `PCKCertChain` type
- Adding methods on `report::Body` to get a vector of the struct
- Adding the verify module

Resolves https://github.com/enarx/enarx/issues/917
Resolves https://github.com/enarx/enarx/issues/789

haraldh commented on src/attestation_types/quote/sigdata.rs:
```diff
@@ -31,6 +32,26 @@ impl From<&[u8; 64]> for ECDSAP256Sig {
 }
 
+impl ECDSAP256Sig {
+    /// Returns r component
+    pub fn r(&self) -> &[u8; 32] {
+        &self.r
+    }
+
+    /// Returns s component
+    pub fn s(&self) -> &[u8; 32] {
+        &self.s
+    }
```
`r` and `s` are already `pub` ... why these methods?

lkatalin replied:
Good point, I removed all of these.

haraldh commented on src/attestation_types/quote/sigdata.rs:
```diff
@@ -57,6 +78,56 @@ impl From<&[u8; 64]> for ECDSAPubKey {
 }
 
+impl ECDSAPubKey {
+    /// Returns x coordinate
+    pub fn x(&self) -> &[u8; 32] {
+        &self.x
+    }
+
+    /// Returns y coordinate
+    pub fn y(&self) -> &[u8; 32] {
+        &self.y
+    }
+
```
`x` and `y` are already `pub` ... why these methods?

haraldh commented on src/attestation_types/quote/sigdata.rs:
```diff
@@ -57,6 +78,56 @@ impl From<&[u8; 64]> for ECDSAPubKey {
+    pub root_cert: X509,
+}
+
+impl PckCertChain {
+    /// Returns leaf certificate
+    pub fn leaf_cert(&self) -> &X509 {
+        &self.leaf_cert
+    }
+
+    /// Returns intermediate certificate
+    pub fn intermed_cert(&self) -> &X509 {
+        &self.intermed_cert
+    }
+
+    /// Returns root certificate
+    pub fn root_cert(&self) -> &X509 {
+        &self.root_cert
+    }
+}
```
These fields are also pub.

haraldh commented on src/attestation_types/quote/sigdata.rs:
```diff
@@ -309,7 +311,9 @@ impl SigData {
     /// Retrieve the QE Cert Data
-    pub fn get_qe_cert_data(&self) -> &Vec<u8> {
-        &self.qe_cert_data
+    pub fn qe_cert_data_vec(&self) -> Vec<u8> {
+        let mut v = Vec::new();
+        v.extend(&self.qe_cert_data);
+        v
```
```rust
        self.qe_cert_data.clone()
```

haraldh commented on src/attestation_types/quote/sigdata.rs:
```diff
@@ -316,4 +316,29 @@ impl SigData {
     }
+
+    /// Returns QE Cert Data as a PCK certificate chain, if QE Cert Data is of the appropriate type
+    pub fn qe_cert_data_pckchain(&self) -> Result<PckCertChain, Box<dyn Error>> {
```
```
    pub fn qe_cert_data_pckchain(&self) -> Result<PckCertChain, QuoteError> {
```

haraldh commented on src/attestation_types/quote/sigdata.rs:
```diff
@@ -316,4 +316,29 @@ impl SigData {
+        }
+
+        // Necessary because the formatting embedded in the Quote is incorrect. Fix later.
+        let chain = String::from_utf8(self.qe_cert_data_vec())?
+            .replace("-----END CERTIFICATE-----", "-----END CERTIFICATE-----\n");
+
+        let pck_cert_chain = X509::stack_from_pem(&chain.as_bytes().to_vec()[..])?;
```
```rust
        let chain = String::from_utf8(self.qe_cert_data_vec())
            .map_err(|e| QuoteError(e.to_string()))?
            .replace("-----END CERTIFICATE-----", "-----END CERTIFICATE-----\n");

        let pck_cert_chain = X509::stack_from_pem(&chain.as_bytes().to_vec()[..])
            .map_err(|e| QuoteError(e.to_string()))?;
```

haraldh commented on src/attestation_types/verify/mod.rs:
```diff
@@ -0,0 +1,114 @@
+
+#[test]
+fn verify_sample_v3quote() {
+    assert!(!verify().is_err());
```
```rust
    assert!(verify().is_ok());
```

haraldh commented:
For `tokio` the minimum required version of rust has to be `1.45.0`.

So for the CI to succeed this is also needed:
```
diff --git a/.github/workflows/test.yml b/.github/workflows/test.yml
index 0675505..70cc3cd 100644
--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -18,7 +18,7 @@ jobs:
       fail-fast: false
       matrix:
         toolchain:
-          - 1.41.0
+          - 1.45.0
           - stable
           - beta
           - nightly
```

haraldh commented:
I also would recommend something like
https://github.com/haraldh/sgx/commit/fe697310c5051bbf3e8f2ccbaff2411831bac451

to decouple the test suite from making online requests every time the test runs.

In the end we probably want something like `sgxctl export --full <filename>` to cache the cert chain.

lkatalin commented:
> I also would recommend something like
> [haraldh@fe69731](https://github.com/haraldh/sgx/commit/fe697310c5051bbf3e8f2ccbaff2411831bac451)
> 
> to decouple the test suite from making online requests every time the test runs.
> 
> In the end we probably want something like `sgxctl export --full <filename>` to cache the cert chain.

@haraldh I've added these changes and added you as co-author. Also opened this issue to cache the cert chain: https://github.com/enarx/sgx/issues/18

lkatalin commented:
@haraldh Thanks so much for these suggestions, they were extremely helpful. :+1: :+1: :+1:

haraldh commented on src/attestation_types/verify/mod.rs:
```diff
@@ -0,0 +1,129 @@
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use sample_v3quote::{SAMPLE_INTEL_CERT_CHAIN, SAMPLE_V3QUOTE};
+
```
still needs:
```diff
diff --git a/src/attestation_types/verify/mod.rs b/src/attestation_types/verify/mod.rs
index 6b72593..bda2b24 100644
--- a/src/attestation_types/verify/mod.rs
+++ b/src/attestation_types/verify/mod.rs
@@ -114,7 +114,10 @@ pub fn verify(quote_bytes: &[u8], trusted_public_pck_chain: &str) -> Result<(),
 mod test {
     use super::*;
 
-    use sample_v3quote::{SAMPLE_INTEL_CERT_CHAIN, SAMPLE_V3QUOTE};
+    use sample_v3quote::SAMPLE_V3QUOTE;
+
+    #[cfg(not(feature = "chain_get"))]
+    use sample_v3quote::SAMPLE_INTEL_CERT_CHAIN;
 
     #[test]
     fn verify_sample_v3quote() {
```

lkatalin replied:
Fixed!

lkatalin commented:
Changing to draft just because I want to include a few more tests for negative cases.

lkatalin commented:
@haraldh I only made changes in a338d6641b30f3a8c225e85ae4472f8c6b577eaf .

haraldh commented on src/attestation_types/verify/cert_chain.rs:
```diff
@@ -0,0 +1,117 @@
+        // chain. If the chain cannot be verified, the leaf will not be
+        // verified.
+        match context.init(&store, &self.leaf, &chain, |c| c.verify_cert()) {
+            Ok(success) => {
+                if success {
+                    Ok(())
+                } else {
+                    Err(VerifyError(format!(
+                        "invalid signature in certificate chain: {}",
+                        context.error()
+                    )))
+                }
+            }
```
```suggestion
            Ok(true) => Ok(()),
            Ok(false) => Err(VerifyError(format!(
                "invalid signature in certificate chain: {}",
                context.error()
            ))),
```

lkatalin replied:
You're right, this is better.

haraldh commented on src/attestation_types/verify/key.rs:
```diff
@@ -0,0 +1,171 @@
+        let mut verifier = Verifier::new(MessageDigest::sha256(), &self.pubkey)?;
+        verifier.update(signed)?;
+        match verifier.verify(sig) {
+            Ok(success) => {
+                if success {
+                    Ok(())
+                } else {
+                    Err(VerifyError("signature verification failed".to_string()))
+                }
+            }
+            Err(_) => Err(VerifyError(
+                "signature validity could not be determined".to_string(),
+            )),
```
```suggestion
            Ok(true) => Ok(()),
            Ok(false) => Err(VerifyError("signature verification failed".to_string())),
            Err(e) => Err(VerifyError(format!(
                "signature validity could not be determined: {}",
                e
            ))),
```

connorkuehl commented on src/attestation_types/verify/key.rs:
```diff
@@ -0,0 +1,167 @@
+pub struct HashError;
+
+impl Error for HashError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        // Generic error, underlying cause isn't tracked.
+        None
+    }
```
```suggestion
```

This is already `std::error::Error`'s default implementation if an implementor does not supply their own.

lkatalin replied:
You're right. Actually I'm removing this whole error because I made a more general one with `VerifyError`. This is one of those things that was left in the code from the other demo.

connorkuehl commented on src/attestation_types/verify/key.rs:
```diff
@@ -0,0 +1,167 @@
+
+    /// DHKE deriving shared secret between self's private key and peer's public key.
+    pub fn derive_shared_secret(&self, peer_key: &PKey<Public>) -> Result<Vec<u8>, Box<dyn Error>> {
+        let ec_priv_key = self.privkey.as_ref().unwrap();
```
Is this `unwrap()` safe or should it be followed by a `.ok_or(...)?;`

connorkuehl commented on src/attestation_types/verify/key.rs:
```diff
@@ -0,0 +1,167 @@
+    /// This creates a new Key from existing PKey value.
+    pub fn new_from_pubkey(pkey: PKey<Public>) -> Self {
+        Key {
+            curve: EcGroup::from_curve_name(Nid::X9_62_PRIME256V1).unwrap(),
```
Is this unwrap safe?

lkatalin replied:
It wasn't. Thanks.

lkatalin commented on src/attestation_types/verify/cert_chain.rs:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    /// This checks that the CertChain's length is under the maximum allowed.
+    pub fn len_ok(&self) -> Result<(), Box<dyn Error>> {
+        if self.chain.len() > self.max_len {
+            panic!("Certificate chain length exceeds max allowable.");
+        }
+        Ok(())
+    }
```
Ah, you're right. I think this should error out because of the kind of check it's meant to be (something we decided on last year about a max allowable chain length), but it shouldn't panic.

lkatalin commented:
Thanks for the good suggestions, @connorkuehl ! I reworked some of these functions and their error handling.
